policy/volatile_ttl.cpp

```diff
@@ -0,0 +1,170 @@
+// Copyright (C) 2018 Bluzelle
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License, version 3,
+// as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include <policy/volatile_ttl.hpp>
+#include <boost/src/boost/include/boost/lexical_cast.hpp>
+
+namespace
+{
+    // TODO: This is a duplicate of code in crud, it would be a good idea to find a way to dry this up.
+    const std::string TTL_UUID{"TTL"};
+}
+
+
+namespace bzn::policy
+{
+    struct ttl_item {std::string key; size_t value_size; size_t ttl;};
+
+    std::set<bzn::key_t>
+    volatile_ttl::keys_to_evict(const database_msg &request, size_t max_db_size)
+    {
+        std::set<bzn::key_t> keys_to_evict;
+
+        auto db_uuid{request.header().db_uuid()};
+        auto key_to_ignore{request.has_update() ? request.update().key() : "" };
+
+        // get all the TTL's in storage
+        auto all_ttls{this->storage->get_keys(TTL_UUID)};
+
+        // get the TTL's associated with our database
+        // filter out the ones for this database, if update, ignore current key
+        auto our_ttls{this->filter_ttls(all_ttls, db_uuid, key_to_ignore)};
+        if (our_ttls.empty())
+        {
+            return keys_to_evict;
+        }
+
+        // Create a vector of [key, value size, ttl]
+        std::vector<ttl_item> ttl_items{get_ttl_items(db_uuid, our_ttls)};
+
+        // 3: sort the keys in order of increasing ttl, secondarily sort on decreasing
+        // key value size
+        std::sort(
+                ttl_items.begin(), ttl_items.end()
```

Comment:
move ttl* up to start of line and lambda on it's own indented 4 spaces

Reply:
Done